scarb/src/ops/proc_macro_server/connection.rs

```diff
@@ -0,0 +1,101 @@
+use std::io::{BufRead, Write};
+use std::thread::JoinHandle;
+
+use crossbeam_channel::{Receiver, Sender};
+use scarb_proc_macro_server_types::jsonrpc::{RpcRequest, RpcResponse};
+use tracing::error;
+
+pub struct Connection {
+    pub sender: Sender<RpcResponse>,
+    pub receiver: Receiver<RpcRequest>,
+    io_threads: IoThreads,
+}
+
+impl Connection {
+    pub fn new() -> Self {
+        let (reader_sender, reader_receiver) = crossbeam_channel::bounded(0);
+        let (writer_sender, writer_receiver) = crossbeam_channel::bounded(0);
```

Comment:
you can use `std::sync::mpsc` here as well, it's baked by crossbeam under the hood anyway, yet less deps to think about

Reply:
I think it is different implementation https://github.com/crossbeam-rs/crossbeam/blob/10f0296a274f5c4b904a9e6bf95b0ca7e37708cb/crossbeam-channel/README.md?plain=1#L16

Reply:
https://www.reddit.com/r/rust/comments/18z6bcz/comment/kgfkgne/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

Reply:
Good to know, but after looking at our case we need `std::sync::mpmc` (we have multiple consumers and single producer for channel that is reading input and sending it to worker threads) not `std::sync::mpsc` and `std::sync::mpmc` is nightly only now.